flecsi/exec/hpx/policy.hh

```diff
@@ -0,0 +1,171 @@
+   //       ///  //////   //////  ////////  //
+
+   Copyright (c) 2016, Triad National Security, LLC
+   All rights reserved.
+                                                                              */
+#pragma once
+
+/*! @file */
+
+#include <flecsi-config.h>
+
+#if !defined(FLECSI_ENABLE_HPX)
+#error FLECSI_ENABLE_HPX not defined! This file depends on HPX!
+#endif
+
+#include <hpx/modules/collectives.hpp>
+
+#include "flecsi/exec/buffers.hh"
+#include "flecsi/exec/hpx/future.hh"
+#include "flecsi/exec/launch.hh"
+#include "flecsi/exec/prolog.hh"
+#include "flecsi/flog.hh"
+#include "flecsi/util/function_traits.hh"
+
+#include <type_traits>
+#include <utility> // forward
+
+namespace flecsi {
+namespace exec {
+namespace detail {
+
+// AA is what the user gives us when calling execute(), PP is what
+// the user defined function/task expects. PP may not be the same
+// as AA, for example, user pass a field_reference as an argument
+// to execute() but the task expects an data accessor as its formal
+// parameter. In this case replace_argument replaces a
+// field_reference with an accessor. This is done through various
+// specialization of the exec::detail::task_param<> template.
+template<class... PP, class... AA>
+auto
+replace_arguments(std::tuple<PP...> * /* to deduce PP */, AA &&... aa) {
+  // Specify the template arguments explicitly to produce references to
+  // unchanged arguments.
+  return std::tuple<decltype(exec::replace_argument<PP>(std::forward<AA>(
+    aa)))...>(exec::replace_argument<PP>(std::forward<AA>(aa))...);
+}
+
+template<typename R>
+struct reduction_helper {
+
+  template<typename T>
+  auto operator()(T const & lhs, T const & rhs) const {
+    return R::combine(lhs, rhs);
+  }
+};
+
+} // namespace detail
+
+template<auto & F, class Reduction, TaskAttributes Attributes, typename... Args>
+auto
+reduce_internal(Args &&... args) {
+  using Traits = util::function_traits<decltype(F)>;
+  using R = typename Traits::return_type;
+
+  // replace arguments in args, for example, field_reference -> accessor.
+  auto params = exec::detail::replace_arguments(
+    static_cast<typename Traits::arguments_type *>(nullptr),
+    std::forward<Args>(args)...);
+
+  // TIP: param_buffers is an RAII type. We create an instance and give
+  // the object a reference to the parameters and name of the task. We then
+  // assign it to the `finalize` variable so it is not destroyed immediately.
+  // The object will be destroyed when reduce_internal() returns. The
+  // ~param_buffers() will then perform the necessary clean up on the
+  // parameters (mostly calling mutator.commit()).
+  auto task_name = util::symbol<F>();
+  auto finalize = param_buffers{params, task_name};
+
+  // Now we have accessors, we need to bind the accessor to real memory
+  // for the data field. We also need to patch up default conversion
+  // from args to params, especially for the future<>.
+  prolog _(params, args...);
```

Comment:
Why was this given a name?  Is its lifetime important?  It should probably get a comment.  (We've considered naming it for the MPI backend, but that's because we were going to use some asynchronous communication and store the completion tokens in the `prolog`.)

Reply:
For now, the name is not needed. I expect however to need to store completion tokens in the future, so I decided to give it a name. If it turns out not to be necessary I will change it back.

Reply:
Will do.